New Ethertype protocol for Arista: Arista Vendor Specific Protocol

nikhilap1 opened the pull request:
Arista applied for, and received the assignment of, a vendor-specific EtherType Protocol Identifier in May of 2016. Details below:

Ethertype number is: D28B
Issue date is: May 12, 2016

Arista Subtype 0x01 is Timestamp L2 Header

The timestamp L2 header consist of the following fields:

Arista EtherType (0xD28B)
    Two-byte protocol subtype of 0x1
    Two-byte protocol version: 0x10 for 64-bit timestamp and 0x20 for 48-bit timestamp
    UTC timestamp value in IEEE 1588 time of day format (either 64-bit or 48-bit) with the lower 32-bits representing nanoseconds and upper bits representing seconds.

Sample tcpdump output:

```
 09:49:13.540584 ethertype Arista Vendor Specific Protocol (0xd28b), length 108: SubType: 0x1, Version: 0x20, Timestamp (48-bit): Seconds 31513, Nanoseconds 565572367, 
 09:49:23.197079 ethertype Arista Vendor Specific Protocol (0xd28b), length 110: SubType: 0x1, Version: 0x10, Timestamp TAI(64-bit): Seconds: 1519811363, Nanoseconds: 258347943, 
```

mcr commented:
Please include a pcap file in tests/ with a sample and test case.

nikhilap1 commented:
@mcr, Included the pcap file and the sample output

niks3089 commented:
Hey guys, any update on when this can be reviewed and merged upstream?

infrastation commented:
The change looks mostly OK, though there are some details that need careful proof-reading and possibly fixing.

infrastation commented:
It may make more sense to omit "AVSP" from the source code as it just unnecessarily repeats it is an Arista-specific protocol without adding meaningful information. Any subsequent Arista-specific protocols would likely end up in `print-arista.c` as well, as it is with the Juniper-specific code in `print-juniper.c`. Additional comments are welcome.

Also this change does not include the changes to `tests/TESTLIST`, which means the new test case is not run during `make check`, this needs to be fixed.

nikhilap1 commented:
Changes done in the last commit

1. Removed "avsp" from the source code
2. Updated 64 bit version to print in new UTC format
3. Added test-case in TESTLIST
4. Resolved merge conflicts.

@infrastation: Please review the changes and let us know if this looks ok

fxlb commented:
Please **rebase** your branch on the-tcpdump-group:master.
There should be no "Merge remote-tracking branch 'original/master' into avsp" in your PR.

fxlb commented:
Use the new GET_ macros instead of EXTRACT_ ones (See 176e182 and ee68aa3).

nikhilap1 commented:
@fxlb: I changed to GET_*.  I also updated the *.out files and the pcap file. 
I tried to re-base my branch to get rid of the "Merge remote-tracking branch" commit but that resulted in merge conflicts when it was trying to re-base around 450 commits. (The merge conflicts were not related to my changes). If that particular commit is not acceptable, I can create another PR

fxlb commented:
Pehaps try : `git reset --hard HEAD~2` before rebase and update.

infrastation commented:
A hard reset will delete commits, in case those happen to be not the commits you want to delete, it will give you a peace of mind if you make a backup of the whole project before doing that.

fxlb commented:
> A hard reset will delete commits

Only the bad merge and the revert.

fxlb commented:
> it will give you a peace of mind if you make a backup of the whole project before doing that.

Even for peace of mind, a backup of the whole project is not necessary.
Your can do:
`git co a-branch; git co -b a-branch-save; git co -`

fxlb commented:
> Perhaps try : `git reset --hard HEAD~2` before rebase and update.

Forget this, Bad idea.

nikhilap1 commented:
I tried other alternatives, nothing seems to work. Should I create a new PR?

fxlb commented:
> Merging this PR shouldn't wouldn't show the bad commit based on my previous such experience though.

The "Merge remote-tracking branch 'original/master' into avsp" make a not very readable git history: 
```
a0b1f024 Address some review comments
0ba34d21 Merge remote-tracking branch 'original/master' into avsp
...
[451 commits]
...
5e7c1a31 New Ethertype protocol for Arista: Arista Vendor Specific Protocol
```

> Should I create a new PR?

I don't think a new PR is necessary.
This should work:
```
# Save your current branch 
git branch -m avsp avsp-save

git checkout master

git pull https://github.com/the-tcpdump-group/tcpdump.git master

git checkout -b avsp

# Get the needed files from avsp-save with commands like:
git show avsp-save:print-arista.c > print-arista.c

# Do the updates/commit(s)

git push -f
```

fxlb commented:
>  tried other alternatives, nothing seems to work

@nikhilap1
Simpler solution now available, see my previous comment.

fxlb commented:
Warnings with clang-7:
```
./print-arista.c:38:7: warning: unused variable 'buf' [-Wunused-variable]
 char buf[128];
      ^
./print-arista.c:22:73: warning: unused parameter 'len' [-Wunused-parameter]
arista_print_ethertype(netdissect_options *ndo, const u_char *bp, u_int len)
                                                                        ^
```

nikhilap1 commented:
Thanks @fxlb. Shows just a single commit now. I fixed the warnings as well

fxlb commented:
Good, the merge is near. Just a few comments.

fxlb commented:
The new GET_ macros do the bounds checks with a setjmp/longjmp logic. Example:
```
/* part of extract.h */

static inline uint16_t
get_be_u_2(netdissect_options *ndo, const u_char *p)
{
        if (!ND_TTEST_2(p))
                longjmp(ndo->ndo_truncated, 1);
        return EXTRACT_BE_U_2(p);
}

[...]

#define GET_BE_U_2(p) get_be_u_2(ndo, (const u_char *)(p))
```
Thus you can remove the ND_TCHECK_ macros (print-arista.c, lines 29, 45, 52).
And you have to add in print-arista.c, line 29:
`ndo->ndo_protocol = "arista";`

nikhilap1 commented:
@fxlb: Done.

nikhilap1 commented on print-arista.c:
```diff
@@ -0,0 +1,70 @@
+
+		}
+	} else {
+		goto trunc;
```
Ok, I've removed trunc label and I am simply returning -1. I hope this is what is required

fxlb commented:
Merged, Thanks for your patience.
